src/execution/window_executor.cpp

```diff
@@ -602,6 +605,143 @@ void WindowExecutorBoundsState::UpdateBounds(idx_t row_idx, DataChunk &input_chu
 	state.Bounds(bounds, row_idx, range, count, boundary_start, boundary_end, partition_mask, order_mask);
 }
 
+//===--------------------------------------------------------------------===//
+// ExclusionFilter
+//===--------------------------------------------------------------------===//
+
+//! Handles window exclusion by piggybacking on the filtering logic.
+//! (needed for first_value, last_value, nth_value)
+class ExclusionFilter {
+public:
+	ExclusionFilter(WindowExclusion exclude_mode_p, idx_t total_count, const ValidityMask &src) : mode(exclude_mode_p) {
+		mask.Initialize(total_count);
+
+		// initialize mask_src:
+		if (src.GetData()) {
+			mask_src = &src;
+		} else {
+			// if src has no data (= represents only one entries), set the mask source to all_ones_mask)
+			one_bits.resize(ValidityMask::ValidityMaskSize(total_count), ~0);
+			all_ones_mask.Initialize(one_bits.data());
+			mask_src = &all_ones_mask;
+		}
+		// copy the data from mask_src
+		FetchFromSource(0, total_count - 1);
+	}
+
+	//! Copy the entries from mask_src to mask, in the index range [begin, end)
+	void FetchFromSource(idx_t begin, idx_t end);
+	//! Apply the current exclusion to the validity mask
+	//! (offset is the current row's index within the chunk)
+	void ApplyExclusion(DataChunk &bounds, idx_t row_idx, idx_t offset);
+	//! Reset the validity mask to match mask_src
+	//! (offset is the current row's index within the chunk)
+	void ResetMask(idx_t row_idx, idx_t offset);
+
+	//! The current peer group's begin
+	idx_t curr_peer_begin;
+	//! The current peer group's end
+	idx_t curr_peer_end;
+	//! The window exclusion mode
+	WindowExclusion mode;
+	//! The validity mask representing the exclusion
+	ValidityMask mask;
+	//! The validity mask upon which mask is based
+	const ValidityMask *mask_src;
+	//! A validity mask consisting of only one entries (needed if no ignore_nulls mask is supplied)
+	ValidityMask all_ones_mask;
+	//! A field of one bits
+	vector<validity_t> one_bits;
```

Comment:
Why can't this be a `ValidityMask` as well?

Reply:
It's only the raw data used by `all_ones_mask`, not a mask of its own (see l. 625). I remember adding it as an attribute for allocation reasons, but maybe there's another way.

Reply:
This mask isn't needed at all. `GetValidityEntry` does the right thing if everything is valid.